### DISCUSSION TITLE
ESP32-C3 but with DIO flash instead of QIO

### DISCUSSION THREAD

**Discussion by godefroi** [microsoft/devicescript#592] [Q&A] [answered]
I have (sourced from Amazon) a very affordable little ESP32-C3 development board. Because the ESP32-C3 is quite IO-poor, the designers of this board have chosen to use the flash in DIO mode instead of QIO mode. This makes the flash slower, but frees up two GPIO pins (GPIO12 and GPIO13) that would normally not be available. This DIO mode *must* be specified when the board is flashed.

Using, for example, .NET Nanoframework, this is specified like this:

`nanoff --platform esp32 --serialport COM8 --flashmode dio --update`

Using Arduino, it looks like this:

`arduino-cli burn-bootloader --fqbn esp32:esp32:esp32c3 --programmer esptool --port COM5 --board-options FlashMode=dio`

For ESPHome, the DIO mode is specified in the .yaml:

```yaml
esphome:
  name: my-esphome-device
  platformio_options:
    board_build.flash_mode: dio
```

Is there a way to specify this in DeviceScript, or am I limited to the "standard" boards that use QIO flash mode?

**Comment by mmoskal** [comment]
We should be able to add an option to devs flash command. However underneath it just invokes esptool - i think you can see what exactly with --verbose. You can just run it yourself directly adding appreciate flash mode option.

**Reply by godefroi** [reply]
I think adding an option to the boards yaml would be the perfect implementation; I don't know how common this setup is, though, so I don't know whether it should be any sort of priority. I could make an attempt at a PR, if that's the sort of thing you folks would be interested in. I'm probably not in a position to exhaustively test the thing, though.

**Reply by mmoskal** [reply]
Sounds good, PR would great!

**Reply by pelikhan** [reply]
Maybe a generic option to add random flags to the esptool invocation?

**Comment by godefroi** [comment, marked as answer]
I have created a (n extremely simple) draft PR here: #606 

I would appreciate any suggestions, or I can mark it ready if you're interested in merging it.

**Reply by pelikhan** [reply]
Your changes have been merged!

**Reply by pelikhan** [reply]
See https://microsoft.github.io/devicescript/devices/esp32#custom-write_flash-argument

**Comment by pelikhan** [comment]
The board.json file has a field to pass flash tool arguments -> https://microsoft.github.io/devicescript/devices/esp32#custom-write_flash-argument
